storage/internal/benchmarks/client_pool.go

```diff
@@ -31,130 +30,125 @@ import (
 	"google.golang.org/grpc"
 )
 
-// clientPool functions much like a sync Pool (https://pkg.go.dev/sync#Pool),
-// except it does not automatically remove items stored in the clientPool.
-// Re-using the clients rather than creating a new one each time reduces overhead
-// (such as re-creating the underlying HTTP client and opening credential files),
-// and is the intended way to use Storage clients.
-//
-// There is no limit to how many clients will be created, but it should be around
-// the order of 5 * min(workers, max_samples).
+// clientPool pools a number of Storage clients for use without blocking, ie.
+// a client that is received through Get may still be in use by one or more other
+// calls.
 type clientPool struct {
-	New     func() *storage.Client
-	clients []*storage.Client
+	clients     []*storage.Client
+	clientQueue chan *storage.Client
 }
 
-func (p *clientPool) Get() *storage.Client {
-	// Create the slice if not already created
-	if p.clients == nil {
-		p.clients = make([]*storage.Client, 0)
+// newClientPool initializes the pool with numClients clients initialized using
+// the initializeClient func.
+// Returns the client pool and a cleanup func to be called to close the pool.
+func newClientPool(initializeClient func() (*storage.Client, error), numClients int) (*clientPool, func()) {
+	p := &clientPool{
+		clients:     make([]*storage.Client, numClients),
+		clientQueue: make(chan *storage.Client, numClients),
 	}
 
-	// If there is an unused client, return it
-	if len(p.clients) > 0 {
-		c := p.clients[0]
-		p.clients = p.clients[1:]
-		return c
-	}
+	for i := 0; i < numClients; i++ {
+		var err error
+		p.clients[i], err = initializeClient()
+		if err != nil {
+			log.Fatalf("initializeClient: %v", err)
+		}
 
-	// Otherwise, create a new client and return it
-	return p.New()
-}
+		// Fill the queue with clients as they are created
+		p.clientQueue <- p.clients[i]
+	}
 
-func (p *clientPool) Put(c *storage.Client) {
-	p.clients = append(p.clients, c)
+	return p, func() {
+		for _, c := range p.clients {
+			c.Close()
+		}
+	}
 }
 
-// we can share clients as long as the app buffer sizes are constant
-var httpClients, gRPCClients *clientPool
+// Get rotates through clients. This means the work may not be evenly distributed,
```

Comment:
Did we consider tracking how many ops are currently running on a client and implementing a priority queue based on that? Not something we have to do now necessarily but might be interesting if we see evidence of a workload being distributed unevenly.

Reply:
Yes, it is something that could be done in the future. 

I considered keeping track of the number of bytes being processed currently on each client and assigning based on that, but I think it may over-complicate the code. 

I did do some basic tests and the clients processed similar (or equal) number of bytes.

Reply:
Sounds good, yeah I think we can leave as-is for now